plugins/inputs/firehose/firehose_request.go

```diff
@@ -0,0 +1,140 @@
+package firehose
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/internal"
+)
+
+type record struct {
+	EncodedData string `json:"data"`
+}
+
+type requestBody struct {
+	RequestID string   `json:"requestId"`
+	Timestamp int64    `json:"timestamp"`
+	Records   []record `json:"records"`
+}
+
+// https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html#requestformat
+type request struct {
+	req  *http.Request
+	body requestBody
+	res  response
+}
+
+// https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html#responseformat
+type response struct {
+	body       responseBody
+	statusCode int
+}
+
+type responseBody struct {
+	RequestID    string `json:"requestId"`
+	Timestamp    int64  `json:"timestamp"`
+	ErrorMessage string `json:"errorMessage,omitempty"`
+}
+
+func (r *request) authenticate(expected config.Secret) error {
+	if expected.Empty() {
+		return nil
+	}
+	key := r.req.Header.Get("x-amz-firehose-access-key")
+	match, err := expected.EqualTo([]byte(key))
+	if err != nil {
+		return fmt.Errorf("comparing keys failed: %w", err)
+	}
+	if !match {
+		r.res.statusCode = http.StatusUnauthorized
+		return fmt.Errorf("unauthorized request %s from %v", r.req.Header.Get("x-amz-firehose-request-id"), r.req.RemoteAddr)
+	}
+	return nil
+}
+
+func (r *request) validate() error {
+	requestID := r.req.Header.Get("x-amz-firehose-request-id")
+	if requestID == "" {
+		r.res.statusCode = http.StatusBadRequest
+		return errors.New("x-amz-firehose-request-id header is not set")
+	}
+
+	// The maximum body size can be up to a maximum of 64 MiB.
+	// https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html
+	if r.req.ContentLength > int64(64*1024*1024) {
+		r.res.statusCode = http.StatusRequestEntityTooLarge
+		return fmt.Errorf("content length too large in request %s", requestID)
```

Comment:
I would add the request ID when printing the error of `validate`, i.e. one caller level up and remove it from the error messages here. What do you think? It's not a must and mostly cosmetic...